Alter authentication

john-westcott-iv opened the pull request:
This change does a couple things:
* Allows a created user to be tied to an Authenticator through AuthenticatorUser on the first login. Previously, if you create user "anna" in the system and then logged in you would end up with the user "anna<hash>".
* Ensures a standard pattern where users are considered unique from all sources. The first source to authenticate a user "rick" gets that username, if another authenticator authenticates a user "rick" (same username) the second user will be "rick<hash>". 
* Better aligns the non-sso login process by adding function docs and documentation around what password authenticators should be doing.
* Adds claims process to local and tacacs authenticators.
* Ensures get_user_name 
* Splits out general authentication utils from the `utils/claims` file.

TODO:
- [ ] Test auth with TACACS (reg and collision)
- [ ] Test auth with LDAP (reg and collision)
- [ ] Test auth with social (reg and collision)
- [x] Address any unit tests and fix coverage (splitting out claims and authentication utils)
- [x] Check user hash creation from social_auth to ensure that it will be consistent the way we are using it

john-westcott-iv committed:
Alter authentication to better allow AuthenticatorUser consumption of existing User

cutwater commented on ansible_base/authentication/models/authenticator_user.py:
```diff
@@ -21,7 +21,7 @@ class AuthenticatorUser(AbstractUserSocialAuth):
 
     @classmethod
-    def create_social_auth(cls, user, uid, slug):
+    def create_social_auth(self, user, uid, slug):
```
Why this change? This is a class method, isn't it?

john-westcott-iv replied:
Thanks, reverted; I think that was an auto-pilot change.

cutwater commented on ansible_base/authentication/utils/claims.py:
```diff
@@ -3,20 +3,18 @@
 
 
-def create_claims(authenticator: Authenticator, username: str, attrs: dict, groups: list) -> (bool, bool, dict, list):
+def create_claims(authenticator: Authenticator, username: str, attrs: dict, groups: list[str]) -> (bool, bool, dict, list):
```
The return type annotation is incorrect. Tuple of types is invalid annotation in general and this function returns dictionary on line 84.

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,120 @@
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = {}, extra_data: dict = {}
+) -> Tuple[AbstractUser, Optional[AuthenticatorUser], bool]:
```
see line 96

```suggestion
) -> Tuple[Optional[AbstractUser], Optional[AuthenticatorUser], Optional[bool]]:
```

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = {}, extra_data: dict = {}
```
Don't use mutable types as a default value.

Python binds default values to parameters at function definition time, which happens once when the function is created and not each time when it's executed.
Therefore it creates implicit global variable that when reused can lead to unexpected side effects and hard to debug bugs:

```
>>> def foo(a, x=[]):
...      x.append(a)
...      print(x)
... 
>>> foo(1)
[1]
>>> foo(2)
[1, 2]
```

General best practice is to avoid usage of mutable values as function defaults.

P.S. There are flake8 and ruff checks that prevent it. I think it will benefit the project if we enable them.

john-westcott-iv replied:
Good catch. Fixing.

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,120 @@
+
+    # ensure the authenticator isn't trying to pass along a cheeky is_superuser in user_details
+    allowed_keys = ["first_name", "last_name", "email"]
+    details = {k: user_details.get(k, "") for k in allowed_keys if k}
```
Why if condition in dict comprehension that iterates over list of constants?
```
 details = {k: user_details.get(k, "") for k in allowed_keys if k}
                                                             ^^^^
```

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,120 @@
+            return None, None, None
+
+    # ensure the authenticator isn't trying to pass along a cheeky is_superuser in user_details
+    allowed_keys = ["first_name", "last_name", "email"]
```
If that's a constant, better to move it to module level and name appropriately

See https://peps.python.org/pep-0008/#constants

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,120 @@
+    from other auth backends.
+    """
+
+    class FakeBackend:
```
Is there a reason to use nested class, which is created on every function call for this function?

john-westcott-iv replied:
This code is actually not new but was split out from ansible_base/authentication/utils/claims.py. I don't know why it was initially implemented this way but we can move it out since its really nothing dynamic.=

john-westcott-iv committed:
Alter expect_log fixture to only count entries that match the input string

john-westcott-iv committed:
Alter expected_log to allow for checking multiple levels of same logger

Before this change you could not do:
   with expect_log(logger, 'debug', 'debug message'):
       with expect_log(logger, 'info', 'info message'):

Because the fixture was mocking logger and then requesting the severity.
So the first call would mock logger and the second call would remock logger.

This alters the code to mock to logger.severity allowing the scenario above to exist.

Note: This would not address attempting to test multiple messages from the same logging level

relrod commented on ansible_base/lib/testing/fixtures.py:
```diff
@@ -130,16 +130,19 @@ def expected_log(no_log_messages):
+
+            call_count = 0
+            for call in logger.call_args_list:
+                if substr in call.args[0]:
```
Hmm, yeah that's a good catch, that if the logger is called with other messages in the same block, we'd fail. :+1: good fix

relrod commented on ansible_base/lib/testing/fixtures.py:
```diff
@@ -130,16 +130,19 @@ def expected_log(no_log_messages):
+            call_count = 0
+            for call in logger.call_args_list:
+                if substr in call.args[0]:
+                    call_count = call_count + 1
```
```suggestion
                    call_count += 1
```

or just replace the whole loop with

```python
call_count = len([call for call in logger.call_args_list if substr in call.args[0]])
```

cutwater replied:
No need to construct a temporary list that is immediately discarded.
```
call_count = sum(1 for call in logger.call_args_list if substr in call.args[0])
```

relrod replied:
Ah, `sum()` with 1 works... I originally left out the list, but count() won't work on a generator so I changed it.

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = dict(), extra_data: dict = dict()
```
Replacing dict literal `{}` with a dict call `dict()` doesn't change anything. The dictionary object is still created at the function definition time. The common pattern in Python for creating mutable function default parameter values is using an immutable default with a condition. For example:

```
def foo(data: Optional[dict] = None):
    if data is None:
        data = {}
```

john-westcott-iv replied:
I don't think I want to make it optional but the `()` were typos. Thanks for catching (again).

cutwater replied:
In this case you should remove the default values. The `user_detauls: dict = dict` is incorrect. This assigns `dict` type as a default value, which I think not what was intended here.

john-westcott-iv replied:
Default removed.

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,116 @@
+            return None, None, None
+
+    # ensure the authenticator isn't trying to pass along a cheeky is_superuser in user_details
+    details = {k: user_details.get(k, "") for k in ["first_name", "last_name", "email"] if k}
```
1. The `if k` condition is still here, seems like a mistake. It's either meant to do something else or it simply can be deleted.
2. Minor(nit): Personally I'd move this list to a constant:

```python
USER_DETAILS_FIELDS = ["first_name", "last_name", "email"] 

# ...

def get_or_create_authenticator_user(...):
    # ...
    details = {k: user_details.get(k, "") for k in USER_DETAILS_FIELDS}
```

john-westcott-iv replied:
I thought about moving it to a constant but we don't use it anywhere else so I was just going to put it inline. 
The goal of this was to make sure we only accept specific fields out of the user_fields instead of users being able to put junk in. I don't remember why the `if` was there but yea, it seems useless at this point in this context.

cutwater commented on ansible_base/authentication/utils/authentication.py:
```diff
@@ -0,0 +1,129 @@
+def determine_username_from_uid_social(**kwargs: dict) -> dict:
+    uid = kwargs.get('details', {}).get('username', None)
+    if not uid:
+        raise AuthException('Unable to get associated username from: {{ kwargs.get("details", None) }}')
```
Looks like an f-string was meant to be used here.

john-westcott-iv replied:
Good catch, this actually should have been a translated string.

cutwater reviewed (approved):
Minor comment, otherwise LGTM
